From 108b49ca305775536cd6d7596265f4c4c56e795c Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 23 Apr 2020 11:37:45 +0300
Subject: [PATCH] [SW-11867] preboot: cmd: i2c slave infra

Required cmd module infrastructure for handling
i2c slave requests from BMC. Support for commads
in u-boot mode like PCI vendor id, temp reading
and device status.

Change-Id: I7e70d1c8e1de8956a49c7d1e36aaed63aaa45d21
---

diff --git a/app/preboot.c b/app/preboot.c
index b2bfbf9..93c6ab9 100644
--- a/app/preboot.c
+++ b/app/preboot.c
@@ -66,6 +66,16 @@
 		.teardown = noop,
 		.run_noreturn = noop,
 	},
+	{	/* I2C SLAVE module */
+		.setup = module_i2c_s_setup,
+		.check = module_i2c_s_check,
+		.disable = module_i2c_s_disable,
+		.run = noop,
+		.enable = module_i2c_s_setup,
+		.run_fast = module_i2c_s_run_fast,
+		.teardown = module_i2c_s_disable,
+		.run_noreturn = noop,
+	},
 };
 
 static void preboot_handle_cmds(void)
diff --git a/common/cmd_module.h b/common/cmd_module.h
index be1dcd2..3a0e26d 100644
--- a/common/cmd_module.h
+++ b/common/cmd_module.h
@@ -85,4 +85,10 @@
 enum cmd_check_res module_uart_check(void);
 void module_uart_run(void);
 
+/* I2C SLAVE */
+void module_i2c_s_setup(void);
+enum cmd_check_res module_i2c_s_check(void);
+void module_i2c_s_disable(void);
+void module_i2c_s_run_fast(void);
+
 #endif /* _HL_CMD_MODULE_H_ */
diff --git a/common/module_i2c_slave.c b/common/module_i2c_slave.c
new file mode 100644
index 0000000..9e4576b
--- /dev/null
+++ b/common/module_i2c_slave.c
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2017-2020 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <stddef.h>
+#include <psoc_global_conf_regs.h>
+#include <psoc_i2c_s_regs.h>
+#include <psoc_i2c_s_masks.h>
+#include <hl_fw_if.h>
+#include <pcie_dbi_regs.h>
+#include <preboot.h>
+
+#include "cmd_module.h"
+#include "uart.h"
+#include "common.h"
+
+#define HL_DEVICE_STATUS_LINUX		0x0
+#define HL_DEVICE_STATUS_PREBOOT	0x1
+
+#define HL_DEVICE_STATUS_CMD_CODE_LEN		0x1
+#define HL_PCIE_VENDOR_ID_CMD_CODE_LEN		0x16
+
+/* TODO: Remove this define after actual temp read is implemented */
+#define DUMMY_MAX_TEMP_MC			39000
+
+static unsigned int cmd;
+static unsigned int read_bytes_count;
+
+enum preboot_i2c_slave_event_id {
+	HL_ASIC_TEMP_CMD_CODE = 0,
+	HL_ASIC_TEMP_MAX_VAL = 3,
+	HL_PCIE_VENDOR_ID_VAL_LOW = 9,
+	HL_PCIE_VENDOR_ID_VAL_HIGH = 10,
+	HL_DEVICE_STATUS_CMD_CODE = 165,
+	HL_DEVICE_STATUS = 166,
+};
+
+static void hl_i2c_slave_init(void)
+{
+	/* Check if already enabled */
+	if (regmap_readl(mmPSOC_I2C_S_IC_ENABLE)) {
+		preboot_puts("I2C SLAVE already initialized\n");
+		return;
+	}
+
+	/* Program IC_SAR with address to which slave responds */
+	regmap_writel(I2C_SLAVE_ADDRESS, mmPSOC_I2C_S_IC_SAR);
+
+	/* Enable I2C Slave */
+	regmap_writel(0x1, mmPSOC_GLOBAL_CONF_I2C_SLV);
+
+	/* Enable I2C Slave interrupts */
+	regmap_writel(0x0, mmPSOC_GLOBAL_CONF_I2C_SLV_INTR_MASK);
+
+	/* Program IC_CON */
+	regmap_writel(0x22, mmPSOC_I2C_S_IC_CON);
+
+	/* Write to IC_INTR_MASK to enable all interrupts */
+	regmap_writel(0x7FFF, mmPSOC_I2C_S_IC_INTR_MASK);
+
+	/* Write to IC_RX_TL to set Rx FIFO threshold level */
+	regmap_writel(0x0, mmPSOC_I2C_S_IC_RX_TL);
+
+	/* Write to IC_TX_TL to set Tx FIFO threshold level */
+	regmap_writel(0x7, mmPSOC_I2C_S_IC_TX_TL);
+
+	/* Write 1 to IC_ENABLE to enable DW_apb_i2c */
+	regmap_writel(0x1, mmPSOC_I2C_S_IC_ENABLE);
+
+	preboot_puts("I2C SLAVE initialized\n");
+}
+
+static u32 i2c_dw_read_clear_intrbits_slave(void)
+{
+	u32 stat;
+
+	/*
+	 * The IC_INTR_STAT register just indicates "enabled" interrupts.
+	 * Ths unmasked raw version of interrupt status bits are available
+	 * in the IC_RAW_INTR_STAT register.
+	 *
+	 * That is,
+	 *   stat = regmap_readl(IC_INTR_STAT);
+	 * equals to,
+	 *   stat = regmap_readl(IC_RAW_INTR_STAT) & regmap_readl(IC_INTR_MASK);
+	 *
+	 * The raw version might be useful for debugging purposes.
+	 */
+	stat = regmap_readl(mmPSOC_I2C_S_IC_INTR_STAT);
+
+	/*
+	 * Do not use the IC_CLR_INTR register to clear interrupts, or
+	 * you'll miss some interrupts, triggered during the period from
+	 * regmap_readl(IC_INTR_STAT) to regmap_readl(IC_CLR_INTR).
+	 *
+	 * Instead, use the separately-prepared IC_CLR_* registers.
+	 */
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_TX_ABRT_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_TX_ABRT);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_RX_UNDER_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_RX_UNDER);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_RX_OVER_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_RX_OVER);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_TX_OVER_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_TX_OVER);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_RX_DONE_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_RX_DONE);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_ACTIVITY_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_ACTIVITY);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_STOP_DET_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_STOP_DET);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_START_DET_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_START_DET);
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_GEN_CALL_MASK)
+		regmap_readl(mmPSOC_I2C_S_IC_CLR_GEN_CALL);
+
+	return stat;
+}
+
+void module_i2c_s_setup(void)
+{
+	hl_i2c_slave_init();
+}
+
+enum cmd_check_res module_i2c_s_check(void)
+{
+	enum cmd_check_res res = NOT_FOUND;
+	unsigned int stat;
+
+	i2c_dw_read_clear_intrbits_slave();
+	stat = regmap_readl(mmPSOC_I2C_S_IC_INTR_STAT);
+
+	if ((stat & PSOC_I2C_S_IC_INTR_STAT_R_RD_REQ_MASK) ||
+	    (stat & PSOC_I2C_S_IC_INTR_STAT_R_RX_FULL_MASK)) {
+		preboot_puts("I2C SLAVE: read request received\n");
+		res = FOUND_FAST_RETURN;
+	}
+
+	return res;
+}
+
+void module_i2c_s_disable(void)
+{
+	unsigned int stat;
+
+	stat = regmap_readl(mmPSOC_I2C_S_IC_INTR_STAT);
+
+	/* Clear requests before disabling */
+	if (stat & (PSOC_I2C_S_IC_INTR_STAT_R_RD_REQ_MASK |
+		    PSOC_I2C_S_IC_INTR_STAT_R_RX_FULL_MASK))
+		module_i2c_s_run_fast();
+
+	regmap_writel(0x0, mmPSOC_I2C_S_IC_ENABLE);
+
+	stat = regmap_readl(mmPSOC_I2C_S_IC_INTR_STAT);
+
+	/* Make sure all requests are handled after disable */
+	if (stat & (PSOC_I2C_S_IC_INTR_STAT_R_RD_REQ_MASK |
+		    PSOC_I2C_S_IC_INTR_STAT_R_RX_FULL_MASK))
+		module_i2c_s_run_fast();
+
+	preboot_puts("I2C SLAVE: disabled\n");
+}
+
+static char i2cslave_pcie_vendor_id_cmd(int command_idx, int bytes_count)
+{
+	char ret;
+	unsigned int data = regmap_readl(mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
+
+	switch (command_idx + bytes_count) {
+	case HL_PCIE_VENDOR_ID_VAL_LOW:
+		data &= 0xFFFF;
+		ret = (data & 0xFF00) >> 8;
+		break;
+	case HL_PCIE_VENDOR_ID_VAL_HIGH:
+		data &= 0xFFFF;
+		ret = data & 0xFF;
+		break;
+	default:
+		ret = 0;
+		break;
+	}
+
+	return ret;
+}
+
+/* TODO; Implement this functionality */
+static int hl_ts_get_max_temp(void)
+{
+	return DUMMY_MAX_TEMP_MC;
+}
+
+static int i2cslave_handle_read_req(void)
+{
+	int max_tmp, ret;
+
+	switch (cmd) {
+	case HL_ASIC_TEMP_CMD_CODE:
+	case HL_ASIC_TEMP_MAX_VAL:
+		max_tmp = hl_ts_get_max_temp();
+		/* Convert from mC to C */
+		ret = max_tmp / 1000;
+		break;
+	case HL_DEVICE_STATUS_CMD_CODE:
+		ret = HL_DEVICE_STATUS_CMD_CODE_LEN;
+		break;
+	case HL_DEVICE_STATUS:
+		ret = HL_DEVICE_STATUS_PREBOOT;
+		break;
+	case HL_PCIE_VENDOR_ID_VAL_LOW:
+	case HL_PCIE_VENDOR_ID_VAL_HIGH:
+		ret = i2cslave_pcie_vendor_id_cmd(cmd, read_bytes_count);
+		break;
+	default:
+		/* NA */
+		ret = 0;
+		break;
+	}
+
+	return ret;
+}
+
+void module_i2c_s_run_fast(void)
+{
+	unsigned int stat;
+	int ret;
+
+	stat = regmap_readl(mmPSOC_I2C_S_IC_INTR_STAT);
+
+	if (stat & PSOC_I2C_S_IC_INTR_STAT_R_RD_REQ_MASK) {
+		if (stat & PSOC_I2C_S_IC_INTR_STAT_R_RX_FULL_MASK) {
+			read_bytes_count = 0;
+
+			cmd = regmap_readl(mmPSOC_I2C_S_IC_DATA_CMD);
+
+			/* clear read request */
+			regmap_readl(mmPSOC_I2C_S_IC_CLR_RD_REQ);
+			stat = i2c_dw_read_clear_intrbits_slave();
+		} else {
+			regmap_readl(mmPSOC_I2C_S_IC_CLR_RD_REQ);
+			regmap_readl(mmPSOC_I2C_S_IC_CLR_RX_UNDER);
+			stat = i2c_dw_read_clear_intrbits_slave();
+		}
+
+		ret = i2cslave_handle_read_req();
+		read_bytes_count++;
+		regmap_writel(ret, mmPSOC_I2C_S_IC_DATA_CMD);
+		preboot_puts("I2C SLAVE: read request processed");
+	}
+}
diff --git a/gaudi/common.h b/gaudi/common.h
index a8b3319..202a46e 100644
--- a/gaudi/common.h
+++ b/gaudi/common.h
@@ -33,3 +33,4 @@
 #define mmNIC2_TPC_PLL_BASE		0xD7A000
 
 #define FUSE_WORDS_PER_BANK		0x10
+#define I2C_SLAVE_ADDRESS		0x4f
diff --git a/goya/common.h b/goya/common.h
index d1784b5..0d42b4f 100644
--- a/goya/common.h
+++ b/goya/common.h
@@ -23,3 +23,4 @@
 #define mmTPC_PLL_BASE		0xE01000
 
 #define FUSE_WORDS_PER_BANK	0x10
+#define I2C_SLAVE_ADDRESS	0x4f
diff --git a/goya2/common.h b/goya2/common.h
index 1fbf575..26dfed2 100644
--- a/goya2/common.h
+++ b/goya2/common.h
@@ -19,3 +19,4 @@
 #define mmPSOC_PCI_PLL_BASE		(mmPSOC_PCI_PLL_CTRL_BASE & 0x00FFFFFF)
 
 #define FUSE_WORDS_PER_BANK		0x40
+#define I2C_SLAVE_ADDRESS		0x6A
diff --git a/goya2_arc_gnu/common.h b/goya2_arc_gnu/common.h
index 9efaa13..f62ec56 100644
--- a/goya2_arc_gnu/common.h
+++ b/goya2_arc_gnu/common.h
@@ -19,6 +19,7 @@
 #define PREBOOT_TEXT_BASE	0x8
 
 #define FUSE_WORDS_PER_BANK	0x40
+#define I2C_SLAVE_ADDRESS	0x6A
 
 /*
  * extracts "num" number of bits from position "pos" in the
